test/README.md

```diff
@@ -0,0 +1,84 @@
+Candid test data
+================
+
+This directory contains Candid test data, to be used to check Candid
+implementations for compliance.
+
+This test suite covers
+
+ * parsing Candid service descriptions
+ * given a Candid type, decoding binary Candid data, asserting success or failure
+ * given a Candid type, parsing textual Candid data, asserting success or failure
+ * comparing such decoded/parsed values, using the Host language’s notion of equality.
+
+Because this test data necessarily has to be host-language agnostic, it is not
+possible to
+
+ * assert the decoded value, as it would be host-language specific
+ * cover encoding, as it would require specifying a host-language value,
+   and also because Candid encoding is not canonical
+
+Implementers are advised to extend this test suite with randomzied
+round-tripping tests (host value → candid → host value).
+
+Format
+======
+
+The test suite contains these kind of files:
+
+ * `<name>.good.did`:
+   A Candid service description, with valid candid syntax.
+   Implementations should be able to parse them.
+
+ * `<name>.bad.did`:
+   A Candid service description, with invalid candid syntax.
+   Implementations should reject parsing them.
+
+ * `<name>.test.did`:
+   A set of Candid tests, written in a grammar that extends the Candid type and value grammar:
+
+   The root production is
+   ```
+   <enctest>  ::= <def>;* <test>;*
+   ```
+   to allow any number of type definitions followed by encoding tests with grammar
+   ```
+   <test> ::=
+      | assert <input> : <tuptype> <desc>
+      | assert <input> !: <tuptype> <desc>
+      | assert <input> == <input> : <tuptype> <desc>
+      | assert <input> != <input> : <tuptype> <desc>
+   <input> ::= <text> | blob <text>
+   <desc> ::=  <text>?
+   ```
+   where the four forms assert
+    * that the input can be decoded/parsed at that type
+    * that the input cannot be decoded/parsed at that type
+    * that the two inputs values decode/parse to the same value
+    * that the two inputs values decode/parse to different values.
+
+   The last two forms refer to the host language's equality, and are useful to
+   assert that due to subtyping, certain information is ignored.
+
+   The `<input>` describes input in textual notation (`<text>`) or binary
+   format (`blob <text>`).
+
+   A `<desc>` is an optional description of the test, potentially useful in
+   the test suite output.
+
+A Candid implementation conforms to this test suite if
+
+ * its service parser (if present) accepts all `*.good.did` files and rejects
+   all `*.bad.did` files in this directory.
+ * all tests from all `*.test.did` files succeed.
+
+Candid implementations that do not support parsing service files can ignore the
+`*.good.did` and `*.bad.did` files.
+
+Candid implementations that do not support parsing the textual prepresentation
+can ignore those tests where all inputs are textual. They should treat a test
```

Comment:
Not sure I see the need for this paragraph. Of course, anybody is free to ignore anything, but do you want to encourage not checking the values? If they bother implementing parsing assertions then they can just as well bother implementing parsing of values.

Reply:
Hmm, due to the stage difference, this implication doesn't quite hold. For example Motoko RTS does not implement parsing the textual representation (and should not have to). Likey we will write a tool that converts this test suite into a motoko test file, and _that_ tool would have a parser for annotations.

I included this sentence to make sure that we include 
```
assert blob "DIDL\00\01\7e\00" == "(false)" : (bool) "bool: false";
```
in the test suite and don't have to _also_ include
```
assert blob "DIDL\00\01\7e\00" : (bool) "bool: false";
```
for those who can’t make sense of the textual code, and who might initially think they should just ignore all assertions with textual representation anywhere.

Reply:
> Hmm, due to the stage difference, this implication doesn't quite hold.

Well, yes, that test would be run at a different stage by such a tool, as a kind of self-validation.

Anyway, I see what you're getting at, don't think it's necessary, but don't feel strongly either.